core/src/main/java/io/undertow/server/handlers/GracefulShutdownHandler.java

```diff
@@ -41,25 +42,49 @@
  */
 public class GracefulShutdownHandler implements HttpHandler {
 
-    private volatile boolean shutdown = false;
+    private static final long SHUTDOWN_MASK = 1L << 63;
+    private static final long ACTIVE_COUNT_MASK = (1L << 63) - 1;
+
+    private static final LongUnaryOperator incrementActive = current -> {
+        long incrementedActiveCount = activeCount(current) + 1;
+        return incrementedActiveCount | (current & ~ACTIVE_COUNT_MASK);
+    };
+
+    private static final LongUnaryOperator incrementActiveAndShutdown =
+            incrementActive.andThen(current -> current | SHUTDOWN_MASK);
+
+    private static final LongUnaryOperator decrementActive = current -> {
+        long decrementedActiveCount = activeCount(current) - 1;
+        return decrementedActiveCount | (current & ~ACTIVE_COUNT_MASK);
+    };
+
     private final GracefulShutdownListener listener = new GracefulShutdownListener();
     private final List<ShutdownListener> shutdownListeners = new ArrayList<>();
 
     private final Object lock = new Object();
 
-    private volatile long activeRequests = 0;
-    private static final AtomicLongFieldUpdater<GracefulShutdownHandler> activeRequestsUpdater = AtomicLongFieldUpdater.newUpdater(GracefulShutdownHandler.class, "activeRequests");
+    private volatile long state = 0;
+    private static final AtomicLongFieldUpdater<GracefulShutdownHandler> stateUpdater =
+            AtomicLongFieldUpdater.newUpdater(GracefulShutdownHandler.class, "state");
 
     private final HttpHandler next;
 
     public GracefulShutdownHandler(HttpHandler next) {
         this.next = next;
     }
 
+    private static boolean isShutdown(long state) {
+        return (state & SHUTDOWN_MASK) != 0;
+    }
+
+    private static long activeCount(long state) {
+        return state & ACTIVE_COUNT_MASK;
+    }
+
     @Override
     public void handleRequest(HttpServerExchange exchange) throws Exception {
-        activeRequestsUpdater.incrementAndGet(this);
-        if (shutdown) {
+        long snapshot = stateUpdater.updateAndGet(this, incrementActive);
+        if (isShutdown(snapshot)) {
             decrementRequests();
```

Comment:
This case could be optimized to avoid CAS increment+decrement after the shutdown flag has been set by extracting the LongUnaryOperator into a loop until CAS succeeds similarly to the jboss-threads view executor pattern: https://github.com/jbossas/jboss-threads/blob/a203e13fd739902f850b55a14dc9c77e46a361a6/src/main/java/org/jboss/threads/EnhancedViewExecutor.java#L250-L280

Unclear if it's helpful to optimize this case though, I'm most interested in the hot common case in which the server is actively accepting requests.

e.g.
```java
for (;;) {
    long snapshot = state;
    if (isShutdown(snapshot)) {
        exchange.setStatusCode(StatusCodes.SERVICE_UNAVAILABLE);
        exchange.endExchange();
        return;
    }
    if (stateUpdater.compareAndSet(this, snapshot, snapshot + 1)) {
        exchange.addExchangeCompleteListener(listener);
        next.handleRequest(exchange);
        break;
   }
}
```